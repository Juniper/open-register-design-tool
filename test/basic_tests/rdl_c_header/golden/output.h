///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
//
/* This is an automatically generated header file for example_addr_map address map */
//
/* DO NOT MODIFY THIS FILE! USE 'openrdt' toolchain to modify register map! */
//
/* Generated on: 2021-10-18 */
//
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////

#ifndef __EXAMPLE_ADDR_MAP_REGISTER_MAP__
#define __EXAMPLE_ADDR_MAP_REGISTER_MAP__

/*
 * bits.h
 *
 * Struct and function declarations for dealing with bit assignment.
 */

#ifndef _BITS_H
#define _BITS_H

#define BITS_PER_LONG 32

// ## allows token concatenation
//X = 1 and Y = 10 would return 110
#define __AC(X,Y)	(X##Y)
#define _AC(X,Y)	__AC(X,Y)

#define _UL(x)		(_AC(x, UL))
#define UL(x)		(_UL(x))

#define BIT(nr) (1UL << (nr))
// BIT defines a bit mask for the specified bit number from 0 to whatever fits into an unsigned long
// so BIT(10) should evaluate to decimal 1024 (which is binary 1 left shifted by 10 bits)

#define GENMASK_INPUT_CHECK(h, l) 0

// h is high index, l is low index in a bitfield
// __GENMASK returns 32 bit number with 1s in the h-to-l field
// if h = 4 and l = 1, __GENMASK would return 00000000000000000000000000011110
#define __GENMASK(h, l) \
	(((~UL(0)) - (UL(1) << (l)) + 1) & \
	 (~UL(0) >> (BITS_PER_LONG - 1 - (h))))

#define GENMASK(h, l) \
	(GENMASK_INPUT_CHECK(h, l) + __GENMASK(h, l))

#endif /* _BITS_H */
/* EXAMPLE_ADDR_MAP_REGISTERS memory map */
enum EXAMPLE_ADDR_MAP_REGS {
  DEVICE_ID = 0x0,
  SPI_CONFIG = 0x4,
  SPI_STATUS = 0x8,
};

/* DEVICE_ID registers */
#define FLD_A GENMASK(7, 0)
#define FLD_B GENMASK(15, 8)
#define FLD_C GENMASK(31, 16)

/* SPI_CONFIG registers */
#define REV_MINOR GENMASK(3, 0)
#define REV_MAJOR GENMASK(7, 4)

/* SPI_STATUS registers */
#define MASTER GENMASK(3, 2)
#define MASTER1 GENMASK(1, 0)
#define TRANSFER GENMASK(7, 4)

#endif
