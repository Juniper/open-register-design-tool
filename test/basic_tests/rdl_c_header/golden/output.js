/*   Ordt 190715.01 autogenerated file 
 *   Input: ./rdl_latex/test.rdl
 *   Parms: ./rdl_latex/test.parms
 *   Date: Wed Sep 15 10:17:15 EDT 2021
 */

register_set example_addr_map "Registers for example_addr_map" {

  address = 0x0;
  register_width = 32;
  register r0 "DEVICE_ID" {
    address = 0x0;
    category = { DYNAMIC_CONFIG };
    description = "{
      Register with Device identifier.
    }";
    integer f2[16] "FLD_C" param {
      description = "{
        Description of example field
      }";
      access_mode = READ_WRITE;
      reset = 0xa0b;
    };
    integer f1[8] "FLD_B" param {
      description = "{
        Description of FLD_B
      }";
      access_mode = READ_WRITE;
      reset = 0x4;
    };
    integer f0[8] "FLD_A" param {
      description = "{
        Description of FLD_A
      }";
      access_mode = READ_WRITE;
      reset = 0x2;
    };
  };
  
  register r1 "REVISION_ID" {
    address = 0x4;
    category = { DYNAMIC_CONFIG };
    description = "{
      Register with revision number
    }";
    nop[24];
    integer f1[4] "REV_MAJOR" param {
      description = "{
        Major revision number
      }";
      access_mode = READ_WRITE;
      reset = 0x0;
    };
    integer f0[4] "REV_MINOR" param {
      description = "{
        Minor revision number
      }";
      access_mode = READ_WRITE;
      reset = 0x1;
    };
  };
  
  register_set_size = 0x8;
};

