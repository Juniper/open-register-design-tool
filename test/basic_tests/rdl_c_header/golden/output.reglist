//   Ordt 230719.01 autogenerated file 
//   Input: ./rdl_c_header/test.rdl
//   Parms: ./rdl_c_header/test.parms
//   Date: Thu Jul 20 12:53:13 EDT 2023
//

0x0              size: 32   fields: 3    bits: 32        r0
0x4              size: 32   fields: 2    bits: 8         r1
0x8              size: 32   fields: 3    bits: 8         r2
