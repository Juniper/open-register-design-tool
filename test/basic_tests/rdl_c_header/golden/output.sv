//   Ordt 190715.01 autogenerated file 
//   Input: ./rdl_latex/test.rdl
//   Parms: ./rdl_latex/test.parms
//   Date: Wed Sep 15 10:17:15 EDT 2021
//

//
//---------- module example_addr_map_jrdl_logic
//
module example_addr_map_jrdl_logic
(
  clk,
  reset,
  d2l_r0_w,
  d2l_r0_we,
  d2l_r0_re,
  d2l_r1_w,
  d2l_r1_we,
  d2l_r1_re,

  l2d_r0_r,
  l2d_r1_r,
  l2h_r0_f0_r,
  l2h_r0_f1_r,
  l2h_r0_f2_r,
  l2h_r1_f0_r,
  l2h_r1_f1_r );

  //------- inputs
  input    clk;
  input    reset;
  input     [31:0] d2l_r0_w;
  input    d2l_r0_we;
  input    d2l_r0_re;
  input     [31:0] d2l_r1_w;
  input    d2l_r1_we;
  input    d2l_r1_re;

  //------- outputs
  output     [31:0] l2d_r0_r;
  output     [31:0] l2d_r1_r;
  output     [7:0] l2h_r0_f0_r;
  output     [7:0] l2h_r0_f1_r;
  output     [15:0] l2h_r0_f2_r;
  output     [3:0] l2h_r1_f0_r;
  output     [3:0] l2h_r1_f1_r;


  //------- reg defines
  logic   [7:0] rg_r0_f0;
  logic   [7:0] reg_r0_f0_next;
  logic   [7:0] l2h_r0_f0_r;
  logic   [7:0] rg_r0_f1;
  logic   [7:0] reg_r0_f1_next;
  logic   [7:0] l2h_r0_f1_r;
  logic   [15:0] rg_r0_f2;
  logic   [15:0] reg_r0_f2_next;
  logic   [15:0] l2h_r0_f2_r;
  logic   [31:0] l2d_r0_r;
  logic   [3:0] rg_r1_f0;
  logic   [3:0] reg_r1_f0_next;
  logic   [3:0] l2h_r1_f0_r;
  logic   [3:0] rg_r1_f1;
  logic   [3:0] reg_r1_f1_next;
  logic   [3:0] l2h_r1_f1_r;
  logic   [31:0] l2d_r1_r;
  
  
  //------- combinatorial assigns for r0 (pio read data)
  always_comb begin
    l2d_r0_r [7:0]  = rg_r0_f0;
    l2d_r0_r [15:8]  = rg_r0_f1;
    l2d_r0_r [31:16]  = rg_r0_f2;
  end
  
  //------- combinatorial assigns for r1 (pio read data)
  always_comb begin
    l2d_r1_r = 32'b0;
    l2d_r1_r [3:0]  = rg_r1_f0;
    l2d_r1_r [7:4]  = rg_r1_f1;
  end
  
  //------- combinatorial assigns for r0
  always_comb begin
    reg_r0_f0_next = rg_r0_f0;
    l2h_r0_f0_r = rg_r0_f0;
    reg_r0_f1_next = rg_r0_f1;
    l2h_r0_f1_r = rg_r0_f1;
    reg_r0_f2_next = rg_r0_f2;
    l2h_r0_f2_r = rg_r0_f2;
    if (d2l_r0_we) reg_r0_f0_next = d2l_r0_w [7:0] ;
    if (d2l_r0_we) reg_r0_f1_next = d2l_r0_w [15:8] ;
    if (d2l_r0_we) reg_r0_f2_next = d2l_r0_w [31:16] ;
  end
  
  //------- reg assigns for r0
  always_ff @ (posedge clk) begin
    if (reset) begin
      rg_r0_f0 <= #1 8'h2;
      rg_r0_f1 <= #1 8'h4;
      rg_r0_f2 <= #1 16'ha0b;
    end
    else begin
      rg_r0_f0 <= #1  reg_r0_f0_next;
      rg_r0_f1 <= #1  reg_r0_f1_next;
      rg_r0_f2 <= #1  reg_r0_f2_next;
    end
  end
  
  //------- combinatorial assigns for r1
  always_comb begin
    reg_r1_f0_next = rg_r1_f0;
    l2h_r1_f0_r = rg_r1_f0;
    reg_r1_f1_next = rg_r1_f1;
    l2h_r1_f1_r = rg_r1_f1;
    if (d2l_r1_we) reg_r1_f0_next = d2l_r1_w [3:0] ;
    if (d2l_r1_we) reg_r1_f1_next = d2l_r1_w [7:4] ;
  end
  
  //------- reg assigns for r1
  always_ff @ (posedge clk) begin
    if (reset) begin
      rg_r1_f0 <= #1 4'h1;
      rg_r1_f1 <= #1 4'h0;
    end
    else begin
      rg_r1_f0 <= #1  reg_r1_f0_next;
      rg_r1_f1 <= #1  reg_r1_f1_next;
    end
  end
  
endmodule

//
//---------- module example_addr_map_jrdl_decode
//
module example_addr_map_jrdl_decode
(
  clk,
  reset,
  leaf_dec_wr_data,
  leaf_dec_addr,
  leaf_dec_block_sel,
  leaf_dec_valid,
  leaf_dec_wr_dvld,
  leaf_dec_cycle,
  leaf_dec_wr_width,
  l2d_r0_r,
  l2d_r1_r,

  dec_leaf_rd_data,
  dec_leaf_ack,
  dec_leaf_nack,
  dec_leaf_accept,
  dec_leaf_reject,
  dec_leaf_retry_atomic,
  dec_leaf_data_width,
  d2l_r0_w,
  d2l_r0_we,
  d2l_r0_re,
  d2l_r1_w,
  d2l_r1_we,
  d2l_r1_re );

  //------- inputs
  input    clk;
  input    reset;
  input     [31:0] leaf_dec_wr_data;
  input     [39:0] leaf_dec_addr;
  input    leaf_dec_block_sel;
  input    leaf_dec_valid;
  input    leaf_dec_wr_dvld;
  input     [1:0] leaf_dec_cycle;
  input     [2:0] leaf_dec_wr_width;
  input     [31:0] l2d_r0_r;
  input     [31:0] l2d_r1_r;

  //------- outputs
  output     [31:0] dec_leaf_rd_data;
  output    dec_leaf_ack;
  output    dec_leaf_nack;
  output    dec_leaf_accept;
  output    dec_leaf_reject;
  output    dec_leaf_retry_atomic;
  output     [2:0] dec_leaf_data_width;
  output     [31:0] d2l_r0_w;
  output    d2l_r0_we;
  output    d2l_r0_re;
  output     [31:0] d2l_r1_w;
  output    d2l_r1_we;
  output    d2l_r1_re;


  //------- wire defines
  logic   [31:0] pio_dec_write_data;
  logic  pio_dec_address;
  logic  pio_dec_read;
  logic  pio_dec_write;
  logic   [39:0] block_sel_addr;
  logic  block_sel;
  logic  leaf_dec_valid_active;
  logic  leaf_dec_wr_dvld_active;
  
  //------- reg defines
  logic   [31:0] d2l_r0_w;
  logic  d2l_r0_we;
  logic  d2l_r0_re;
  logic   [31:0] d2l_r1_w;
  logic  d2l_r1_we;
  logic  d2l_r1_re;
  logic  leaf_dec_valid_hld1;
  logic  leaf_dec_valid_hld1_next;
  logic  leaf_dec_wr_dvld_hld1;
  logic  leaf_dec_wr_dvld_hld1_next;
  logic  pio_write_active;
  logic  pio_read_active;
  logic  pio_dec_address_d1;
  logic   [31:0] pio_dec_write_data_d1;
  logic   [31:0] dec_pio_read_data;
  logic   [31:0] dec_pio_read_data_d1;
  logic  dec_pio_ack;
  logic  dec_pio_nack;
  logic  dec_pio_ack_next;
  logic  dec_pio_nack_next;
  logic  pio_internal_ack;
  logic  pio_internal_nack;
  logic  pio_external_ack;
  logic  pio_external_nack;
  logic  pio_external_ack_next;
  logic  pio_external_nack_next;
  logic  pio_no_acks;
  logic  pio_activate_write;
  logic  pio_activate_read;
  logic   [31:0] dec_pio_read_data_next;
  logic  external_transaction_active;
  
  
  //------- assigns
  assign  pio_dec_write_data = leaf_dec_wr_data;
  assign  dec_leaf_rd_data = dec_pio_read_data;
  assign  dec_leaf_ack = dec_pio_ack;
  assign  dec_leaf_nack = dec_pio_nack;
  assign  pio_dec_address = leaf_dec_addr [2] ;
  assign  block_sel_addr = 40'h0;
  assign  block_sel = leaf_dec_block_sel;
  assign  leaf_dec_wr_dvld_active = leaf_dec_wr_dvld | leaf_dec_wr_dvld_hld1;
  assign  leaf_dec_valid_active = leaf_dec_valid | leaf_dec_valid_hld1;
  assign  dec_leaf_accept = leaf_dec_valid & block_sel;
  assign  dec_leaf_reject = leaf_dec_valid & ~block_sel;
  assign  pio_dec_read = block_sel & leaf_dec_valid_active & (leaf_dec_cycle == 2'b10);
  assign  pio_dec_write = block_sel & leaf_dec_wr_dvld_active & (leaf_dec_cycle[1] == 1'b0);
  assign  dec_leaf_retry_atomic = 1'b0;
  assign  dec_leaf_data_width = 3'b0;
  
  //------- combinatorial assigns for pio read data
  always_comb begin
    dec_pio_read_data = dec_pio_read_data_d1;
  end
  
  //------- reg assigns for pio read data
  always_ff @ (posedge clk) begin
    if (reset) begin
      dec_pio_read_data_d1 <= #1  32'b0;
    end
    else begin
      dec_pio_read_data_d1 <= #1 dec_pio_read_data_next;
    end
  end
  
  //------- reg assigns for pio i/f
  always_ff @ (posedge clk) begin
    if (reset) begin
      pio_write_active <= #1  1'b0;
      pio_read_active <= #1  1'b0;
    end
    else begin
      pio_write_active <= #1  pio_write_active ? pio_no_acks : pio_activate_write;
      pio_read_active <= #1  pio_read_active ? pio_no_acks : pio_activate_read;
    end
    pio_dec_address_d1 <= #1   pio_dec_address;
    pio_dec_write_data_d1 <= #1  pio_dec_write_data;
  end
  
  //------- combinatorial assigns for leaf i/f
  always_comb begin
    leaf_dec_valid_hld1_next = leaf_dec_valid | leaf_dec_valid_hld1;
    if (dec_pio_ack_next | dec_pio_nack_next) leaf_dec_valid_hld1_next = 1'b0;
    leaf_dec_wr_dvld_hld1_next = leaf_dec_wr_dvld | leaf_dec_wr_dvld_hld1;
    if (dec_pio_ack_next | dec_pio_nack_next | leaf_dec_valid) leaf_dec_wr_dvld_hld1_next = 1'b0;
  end
  
  //------- reg assigns for leaf i/f
  always_ff @ (posedge clk) begin
    if (reset) begin
      leaf_dec_valid_hld1 <= #1  1'b0;
      leaf_dec_wr_dvld_hld1 <= #1  1'b0;
    end
    else begin
      leaf_dec_valid_hld1 <= #1 leaf_dec_valid_hld1_next;
      leaf_dec_wr_dvld_hld1 <= #1 leaf_dec_wr_dvld_hld1_next;
    end
  end
  
  //------- combinatorial assigns for pio ack/nack
  always_comb begin
    pio_internal_nack = (pio_read_active | pio_write_active) & ~pio_internal_ack & ~external_transaction_active;
    dec_pio_ack_next = (pio_internal_ack | (pio_external_ack_next & external_transaction_active));
    dec_pio_nack_next = (pio_internal_nack | (pio_external_nack_next & external_transaction_active));
    pio_no_acks = ~(dec_pio_ack | dec_pio_nack | pio_external_ack | pio_external_nack);
    pio_activate_write = (pio_dec_write & ~(dec_pio_ack | dec_pio_nack));
    pio_activate_read = (pio_dec_read & ~(dec_pio_ack | dec_pio_nack));
  end
  
  //------- reg assigns for pio ack/nack
  always_ff @ (posedge clk) begin
    if (reset) begin
      dec_pio_ack <= #1 1'b0;
      dec_pio_nack <= #1 1'b0;
      pio_external_ack <= #1  1'b0;
      pio_external_nack <= #1  1'b0;
    end
    else begin
      dec_pio_ack <= #1 dec_pio_ack ? 1'b0 : dec_pio_ack_next;
      dec_pio_nack <= #1 dec_pio_nack ? 1'b0 : dec_pio_nack_next;
      pio_external_ack <= #1 pio_external_ack_next;
      pio_external_nack <= #1 pio_external_nack_next;
    end
  end
  
  
  //------- address decode
  always_comb begin
    pio_internal_ack = 1'b0;
    external_transaction_active = 1'b0;
    pio_external_ack_next = 1'b0;
    pio_external_nack_next = 1'b0;
    dec_pio_read_data_next = 32'b0;
    
    d2l_r0_w = pio_dec_write_data_d1  [31:0] ;
    d2l_r0_we = 1'b0;
    d2l_r0_re = 1'b0;
    d2l_r1_w = pio_dec_write_data_d1  [31:0] ;
    d2l_r1_we = 1'b0;
    d2l_r1_re = 1'b0;
    
    casez(pio_dec_address_d1)
    //  Register: r0     Address: 0x0     External: false
    1'b0:
      begin
        d2l_r0_we = pio_write_active & ~dec_pio_ack;
        d2l_r0_re = pio_read_active & ~dec_pio_ack;
        pio_internal_ack =  pio_read_active | pio_write_active;
        dec_pio_read_data_next  [31:0]  = l2d_r0_r;
      end
    //  Register: r1     Address: 0x4     External: false
    1'b1:
      begin
        d2l_r1_we = pio_write_active & ~dec_pio_ack;
        d2l_r1_re = pio_read_active & ~dec_pio_ack;
        pio_internal_ack =  pio_read_active | pio_write_active;
        dec_pio_read_data_next  [31:0]  = l2d_r1_r;
      end
    endcase
  end
  
endmodule

//
//---------- module example_addr_map_pio
//
module example_addr_map_pio
(
  clk,
  reset,
  leaf_dec_wr_data,
  leaf_dec_addr,
  leaf_dec_block_sel,
  leaf_dec_valid,
  leaf_dec_wr_dvld,
  leaf_dec_cycle,
  leaf_dec_wr_width,

  l2h_r0_f0_r,
  l2h_r0_f1_r,
  l2h_r0_f2_r,
  l2h_r1_f0_r,
  l2h_r1_f1_r,
  dec_leaf_rd_data,
  dec_leaf_ack,
  dec_leaf_nack,
  dec_leaf_accept,
  dec_leaf_reject,
  dec_leaf_retry_atomic,
  dec_leaf_data_width );

  //------- inputs
  input    clk;
  input    reset;
  input     [31:0] leaf_dec_wr_data;
  input     [39:0] leaf_dec_addr;
  input    leaf_dec_block_sel;
  input    leaf_dec_valid;
  input    leaf_dec_wr_dvld;
  input     [1:0] leaf_dec_cycle;
  input     [2:0] leaf_dec_wr_width;

  //------- outputs
  output     [7:0] l2h_r0_f0_r;
  output     [7:0] l2h_r0_f1_r;
  output     [15:0] l2h_r0_f2_r;
  output     [3:0] l2h_r1_f0_r;
  output     [3:0] l2h_r1_f1_r;
  output     [31:0] dec_leaf_rd_data;
  output    dec_leaf_ack;
  output    dec_leaf_nack;
  output    dec_leaf_accept;
  output    dec_leaf_reject;
  output    dec_leaf_retry_atomic;
  output     [2:0] dec_leaf_data_width;


  //------- wire defines
  logic   [31:0] d2l_r0_w;
  logic  d2l_r0_we;
  logic  d2l_r0_re;
  logic   [31:0] d2l_r1_w;
  logic  d2l_r1_we;
  logic  d2l_r1_re;
  logic   [31:0] l2d_r0_r;
  logic   [31:0] l2d_r1_r;
  
  
  example_addr_map_jrdl_decode pio_decode ( .* );
    
  example_addr_map_jrdl_logic pio_logic ( .* );
    
endmodule

