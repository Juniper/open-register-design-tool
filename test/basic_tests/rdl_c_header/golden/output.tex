%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% This file is generated documentation for example_addr_map address map
%
% DO NOT MODIFY THIS FILE! USE 'openrdt' toolchain to modify register map!
%
% Generated on: 2021-09-15
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\gdef\TROPICTEXLIBPATH {\detokenize{../../../modules/ts-latex-lib/tropic_tex_lib}}

\documentclass{\TROPICTEXLIBPATH/tropic_datasheet}

\title{TODO!}
\begin{document}

\onecolumn

% Register Summary table
\begin{TropicRatioTable3Col}
{0.15} {0.65} {0.2}
{Address Offset & Register Name & Reset Value}
0x0 & DEVICE_ID & 0x0A0B0402\Ttlb%
0x4 & REVISION_ID & 0x00000001\Ttlb%
\end{TropicRatioTable3Col}

\pagebreak

\begin{TropicRegisterTable}
  {DEVICE_ID}%
  {0x0}%
  FLD_A & RW & 0x2 & 7:0 & Description of FLD_A \Ttlb
  FLD_B & RW & 0x4 & 15:8 & Description of FLD_B \Ttlb
  FLD_C & RW & 0xA0B & 31:16 & Description of example field \Ttlb
\end{TropicRegisterTable}

\begin{TropicRegisterTable}
  {REVISION_ID}%
  {0x4}%
  REV_MINOR & RW & 0x1 & 3:0 & Minor revision number \Ttlb
  REV_MAJOR & RW & 0x0 & 7:4 & Major revision number \Ttlb
\end{TropicRegisterTable}

\end{document}
